trustmanager/keydbstore.go

```diff
@@ -0,0 +1,153 @@
+package trustmanager
+
+import (
+	"database/sql"
+	"fmt"
+	"sync"
+
+	"github.com/docker/notary/pkg/passphrase"
+	"github.com/endophage/gotuf/data"
+	"github.com/jinzhu/gorm"
+	gojose "github.com/square/go-jose"
+)
+
+// KeyDBStore persists and manages private keys on a SQL database
+type KeyDBStore struct {
+	sync.Mutex
+	db         gorm.DB
+	passphrase string
+	encrypter  gojose.Encrypter
+	cachedKeys map[string]data.PrivateKey
+}
+
+// GormPrivateKey represents a PrivateKey in the database
+type GormPrivateKey struct {
+	gorm.Model
+	KeyID      string `sql:"not null;unique"`
+	Encryption string `sql:"not null"`
+	Algorithm  string `sql:"not null"`
+	Public     []byte `sql:"not null"`
+	Private    string `sql:"not null"`
+}
+
+// TableName sets a specific table name for our GormPrivateKey
+func (g GormPrivateKey) TableName() string {
+	return "private_keys"
+}
+
+// NewKeyDBStore returns a new KeyDBStore backed by a SQL database
+func NewKeyDBStore(passphraseRetriever passphrase.Retriever, dbType string, dbSQL *sql.DB) (*KeyDBStore, error) {
+	cachedKeys := make(map[string]data.PrivateKey)
+
+	// Retreive the passphrase that will be used to encrypt the keys
+	passphrase, _, err := passphraseRetriever("", "", false, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	// Setup our encrypted object
+	encrypter, err := gojose.NewEncrypter(gojose.A256GCMKW, gojose.A256GCM, []byte(passphrase))
+	if err != nil {
+		return nil, err
+	}
+
+	// Open a connection to our database
+	db, _ := gorm.Open(dbType, dbSQL)
+
+	return &KeyDBStore{db: db,
+		passphrase: passphrase,
+		encrypter:  encrypter,
+		cachedKeys: cachedKeys}, nil
+}
+
+// AddKey stores the contents of a private key. Both name and alias are ignored,
+// we always use Key IDs as name, and don't support aliases
+func (s *KeyDBStore) AddKey(name, alias string, privKey data.PrivateKey) error {
+	encryptedKey, err := s.encrypter.Encrypt(privKey.Private())
+	if err != nil {
+		return err
+	}
+
+	// Encrypt the private key material
+	encryptedPrivKeyStr := encryptedKey.FullSerialize()
+
+	gormPrivKey := GormPrivateKey{
+		KeyID:      privKey.ID(),
+		Encryption: string(gojose.PBES2_HS512_A256KW),
```

Comment:
This was here until we realized that gojose didn't implement PBES. Forgot to change it.

Taking it from the encrypter sounds good.